sovryn-token-bridge/federator/src/main.js

```diff
@@ -94,45 +123,66 @@ let pollingInterval = config.runEvery * 1000 * 60; // Minutes
 
 let scheduler = new Scheduler(pollingInterval, logger, { run: () => run() });
 
-startServices().catch(err => { console.error('Error starting services:', err) });
+startServices().catch((err) => {
+    console.error('Error starting services:', err);
+});
 
 async function startServices() {
     let isEth;
     try {
         isEth = await clientId.isEthereumChain();
-        console.log("isEth: " + isEth);
-    } catch(err) {
+        console.log('isEth: ' + isEth);
+    } catch (err) {
         logger.error('Unhandled Error on isEthereumChain()', err);
         process.exit();
-    }   
+    }
 
-    if (isEth){
+    if (isEth) {
         try {
             await gasServices.startGasServices();
-        } catch(err) {
+        } catch (err) {
             logger.error('Cannnot start ETH gas services()', err);
             process.exit();
         }
-    };
+    }
+
+    try {
+        await p2pNode.start();
+    } catch (err) {
+        logger.error("Couldn't start P2P network", err);
+        process.exit();
+    }
 
     scheduler.start().catch((err) => {
         logger.error('Unhandled Error on start()', err);
     });
 }
-    
+
 async function run() {
-    try {
-        console.log("before mainfed");
-        await mainFederator.run();
-        console.log("before sidefed");
-        await sideFederator.run();
-    } catch(err) {
-        logger.error('Unhandled Error on run()', err);
-        process.exit();
+    if (p2pNode.getPeerAmount() < config.minimumPeerAmount) {
+        logger.info('Waiting for enough peers');
+        return;
+    }
+
+    if (p2pNode.isLeader()) {

```

Comment:
I would call the `.run()` methods for all federators, even those that are not leaders, and check the leader status in the `run()` method itself.

But note that there are a couple of gotchas:
- A new node can join the network and become the leader when the `run()` of a previous leader is in process, which can result in `run()` being called from two federators at the same time.
- If the leader stays the same, this will mean that one node will bear the burden of all gas costs. This is probably acceptable but good to keep in mind.
- Only the leader will be processing new logs which means that the other nodes will not update their last processed block, and when the leader is changed, there will be a delay when the new leader starts processing blocks again.

Alternatively the leader could be ditched alltogether and all nodes could process logs and potentially send transactions to the blockchain every iteration, but then it would require some solution to avoid wasting gas by sending the same transaction multiple times.